cpp/src/parquet/arrow/schema.cc

```diff
@@ -989,8 +989,8 @@ Result<bool> ApplyOriginalStorageMetadata(const Field& origin_field,
     // so no need to recurse on value types.
     const auto& dict_origin_type =
         checked_cast<const ::arrow::DictionaryType&>(*origin_type);
-    inferred->field = inferred->field->WithType(
-        ::arrow::dictionary(::arrow::int32(), inferred_type, dict_origin_type.ordered()));
+    inferred->field = inferred->field->WithType(::arrow::dictionary(
```

Comment:
I think this fix makes sense, though it might introduce a benign breaking change to the downstream users.